### PR TITLE
Feat view user programs list and program

### LINKED ISSUE #75: View programs list and program where user is representative
## Description
As a GSoC student,
I need create view programs list and programs from route `MyOrganization` > `Portfolio`,
so that I can view programs list and selected program details where the logged in user is the organization representative.

## Mocks
NA

### Update [Required]
- [ ] View list of programs under `MyOrganization` > `Portfolio
- [ ] View Program profile when selecting a program under the list of programs above

### Enhancement to Update [Optional]
NA

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 day

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Create View programs list and program profile functionalities

Fixes #75 

### Type of Change:
- Code
- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
WIP
- login as a user who has an organization that already has a program (you can create this using BIT backend on Swagger UI as the frontend Create/Update program have not yet completed at this stage)
- go to `My Space` > `My Organization` to view the organization profile
- go to the bottom of the organization profile and click the `Go to Portfolio` button. See the programs list
Screen Shot 2020-08-24 at 1 22 47 am

- select a program from the list and seee the program profile page
Screen Shot 2020-08-24 at 1 23 12 am

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] New and existing unit tests pass locally with my changes

**IMPORTANT** 
To test this PR, you need to run the bit backend server from the branch on [PR#116](https://github.com/anitab-org/bridge-in-tech-backend/pull/116) **ONLY** as this PR has pages with date and timezone fields that are already link to the refactored date fields on backend. (PS: don't forget, although not mentioned, you are **ALWAYS** expected to also run MS backend server using **ONLY** [ms-backend-server](https://github.com/mtreacy002/mentorship-backend/tree/ms-backend-server) with any of BIT Frontend PRs)

### CODE REVIEW

File: src/AuthContext.js
Comment (meenakshi-dhanani) [inline src/AuthContext.js]: This line should not be here

File: src/AuthContext.js
Comment (meenakshi-dhanani) [inline src/AuthContext.js]: this should also not be here

File: src/AuthContext.js
Comment (meenakshi-dhanani) [inline src/AuthContext.js]: I feel like the AuthContext changes from the previous commit aren't reflecting

File: src/Navigation.js
Comment (meenakshi-dhanani) [inline src/Navigation.js]: I removed these two lines in the previous commit. Have we rebased?

**Comment by isabelcosta** [comment]
@mtreacy002 can you please provide an image of the user interface you changed in the pull request?

**Comment by mtreacy002** [comment]
> @mtreacy002 can you please provide an image of the user interface you changed in the pull request?

@isabelcosta. This PR is still a WIP, but I'll edit description and add some images I have so far. Although, this might change later once the PR is done.

**Comment by mtreacy002** [comment]
@meenakshi-dhanani . I've managed to get up to the EditProgram using path MyOrganization > Profile > click button `Go to Portfolio` on the bottom of the page > select Program from list.

I'm still trying to make the fetch with multiple api calls work using path MyOrganization > Portfolio. Let me know if you have any resources that can help with the multiple fetch api calls. If we can't solve this by tomorrow same time tonight, perhaps we might need to remove the `Portfolio` option from thee Navbar option under MyOrganization and let user get the program using only Profile path using just `MyOrganization` as the name (no dropdown selection`. The switch between profile and programs then can be done using a button on each of the pages like I did here. 

Screen Shot 2020-08-24 at 1 22 36 am

**Comment by mtreacy002** [comment]
Update @meenakshi-dhanani . I've modified the programs list page to add the `Add program` button.
* When organization already has a program:
Screen Shot 2020-08-24 at 11 31 31 am

* when organization does not have any program
Screen Shot 2020-08-24 at 11 32 06 am

**Comment by meenakshi-dhanani** [comment]
@mtreacy002 can you check if this PR has the changes of the fix that you pushed yesterday morning? Because for a user, that doesn't have an organization it still doesn't show me the create organization form

**Comment by mtreacy002** [comment]
> @mtreacy002 can you check if this PR has the changes of the fix that you pushed yesterday morning? Because for a user, that doesn't have an organization it still doesn't show me the create organization form

@meenakshi-dhanani . This was originally explained through the error message. However, the additional instruction was removed by your request [here](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#discussion_r469711565). Do you want me to add this back?

**Comment by mtreacy002** [comment]
> @mtreacy002 can you check if this PR has the changes of the fix that you pushed yesterday morning? Because for a user, that doesn't have an organization it still doesn't show me the create organization form

@meenakshi-dhanani , yeah, I forgot pulled the changes to this PR 😁. I have now pulled and pushed it. Can you please check again? Thanks

**Comment by meenakshi-dhanani** [comment]
@mtreacy002 can you fix the conflicts and push again?

**Comment by mtreacy002** [comment]
Update @meenakshi-dhanani . I've pull --rebase the recent merge to develop which resolved the conflict. Can you please re-approve? Thanks.

**Comment by meenakshi-dhanani** [comment]
This PR still breaks for me. When I have an organization, that has no programs, and I click on View programs, it fails

```
TypeError: programs.map is not a function
Programs
src/organizations/Programs.jsx:69
  66 |     <th>Creation Date</th>
  67 |   </tr>
  68 | </thead>
> 69 | <tbody>
     | ^  70 |   {programs.map((program) => (
  71 |     <tr key={program.id}>
  72 |       <td><Link to={{
View compiled
▶ 17 stack frames were collapsed.
(anonymous function)
src/organizations/Programs.jsx:29
  26 | fetch(`${BASE_API}/organizations/${organization.id}/programs`, requestOrganization)
  27 |   .then(async response => {
  28 |     const data = await response.json();
> 29 |     if (response.ok)
     | ^  30 |       return setPrograms(data);
  31 |     return setResponseMessage(data.message);
  32 |   })
View compiled
This screen is visible only in development. It will not appear if the app crashes in production.
```

because the API returns this:

```
{
    "id": 1,
    "username": "meena",
    "name": "Meenakshi",
    "slack_username": null,
    "bio": null,
    "location": null,
    "occupation": null,
    "current_organization": null,
    "interests": null,
    "skills": null,
    "need_mentoring": false,
    "available_to_mentor": true,
    "is_available": null
}
```

**Comment by mtreacy002** [comment]
@meenakshi-dhanani , I still can't understand how you get this bug. Coz I tried it on my end, it gives me the right error message. 

Here's what I've done:
- Login as a user who's representative of an organization but has no programs on the database
I login as Terri White
Screen Shot 2020-08-27 at 9 34 10 am

Terrt white has user ID 3
Screen Shot 2020-08-27 at 9 36 43 am

He has organization with ID 2
Screen Shot 2020-08-27 at 9 34 33 am

This is list of programs in the data (all belong to organization ID 1 which is not of Terri White)
Screen Shot 2020-08-27 at 9 34 42 am

And here is the response I've got on Swagger UI when I ask for list of programs for organization 2 that belongs to Terri White

Screen Shot 2020-08-27 at 9 41 32 am

on the Web UI, I get the same error message (so it didn't break for me)

Screen Shot 2020-08-27 at 9 43 51 am


Can you do few things for me:

1> Make sure you run the develop branch as backend server that already uptodate with the anitaB develop branch (if you're running from your fork develop branch)
2> Make sure you pull the latest code for frontend on this PR
3> Make sure you have dropped all the tables and recreated them since the enum type has changed
4> Follow the steps I made above. Test from Swagger UI first. Make sure you give the correct organization ID to get list of programs, then test on UI
5> If this still fails, send me the screenshots of everything just like I did above
6> we need a catch up session today if you want to make sure this works on your end before approving this PR


CC @ramitsawhney27 , @foongminwong or @isabelcosta , can you also please test this on your end to see if this is also failing for you? If this is working well for you, can you then approve the PR? I can work with @meenakshi-dhanani to help her troubleshoot the bug on her end. **THIS IS IMPORTANT**  since we only have today and tomorrow to finalize this open PRs on frontend. My demo is Saturday and I can only demo whatever in the develop branch. Thanks in advance.

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers , I've just synced and pushed this PR commits to the PR before this (PR #70 ).

**Comment by meenakshi-dhanani** [comment]
@mtreacy002 I am caught up today. Can you clean up your database and do the following steps: 
- Register a user
- Declare he represents an organization
- Create an organization

The backend API /organizations/{id}/programs returns an object for me. Not 404. I have deleted my database and recreated it. There's something with the backend API, it's not about the frontend.

**Comment by mtreacy002** [comment]
Ok, I'll try and let you know.

**Comment by mtreacy002** [comment]
@meenakshi-dhanani . I've tried, I still can't get the same bug. Here's what I've done:
- Created a completely new database named test_pr_76
- connect ms and bit backend to the new db
- register a user from UI, then login, create an organization with no program
- Now, if I click on `Go to Programs` straight after I got the message `Organization has succeffully created` I'll see an error response
Screen Shot 2020-08-27 at 10 16 17 pm
This is because I didn't put in place mechanism to prevent this from happening (a different bug, but we can open a different issue on that)
- Now step away from the page, for example go to Organizations > select the user's organization, then select `View~ under `Programs`. see the expected error message
Screen Shot 2020-08-27 at 10 16 48 pm

- Now go back to `My Space` > My Organization` > `Go to Programs`. see expected error message.
Screen Shot 2020-08-27 at 10 16 57 pm


Here're the screenshots of the new database
- users table
Screen Shot 2020-08-27 at 10 19 15 pm

- organizations table
Screen Shot 2020-08-27 at 10 19 05 pm

- programs table
Screen Shot 2020-08-27 at 10 19 10 pm


From this test, you can see that I still can't replicate the bug that you're facing. Maybe you should try with this method (totally drop and recreate the database, not just dropping the tables.

**Comment by mtreacy002** [comment]
@meenakshi-dhanani , can you please confirm that you have the same 2 commit IDs with the one on this PR as the one running on your local terminal?

on the PR
Screen Shot 2020-08-27 at 11 40 12 pm

the same as on my terminal (top 2 commits)
Screen Shot 2020-08-27 at 11 40 19 pm

**Comment by meenakshi-dhanani** [comment]
ProgramList

So the frontend looks fine. The problem it wasn't working is because of mentorship. This works!

**Review by meenakshi-dhanani — APPROVED** [review]

**Comment by mtreacy002** [comment]
@foongminwong can you please also give your review here? we're so close to wrap this up... I'm excited 😁😁😁

**Comment by foongminwong** [comment]
Hey @mtreacy002 can you fix the conflicting files? 💪

**Comment by mtreacy002** [comment]
> Hey @mtreacy002 can you fix the conflicting files? 💪

@foongminwong , I forgot to put the blocker for these 2 last PR. Actually PR #79 has to be merged after PR #76. But this is not a big issue, since you have tested that PR #79 works as it should be which means this PR code that was carried into PR #79 already got merged as well. I will closed this issue now since I don't want to undo my code that I've added to PR #79 that is not in this PR.

**Comment by mtreacy002** [comment]
@foongminwong . Can you please confirm by posting a comment here some tests on the develop branch (after the PR #79 merged) for the functionalities mentioned on this PR (list programs and view other program). They should work as they should be. If not, we can open another issue as a bug for it.

**Comment by foongminwong** [comment]
Yes, the functionality works after merging PR #79 ✨

test-bit-web-after-pr-79-merge

**Pull request closed without merging**
